rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+  }
+
+  *taken = true;
+  return RMW_RET_OK;
+}
+
+/// SEND SERVICE RESPONSE ======================================================
+// Serialize and publish a ROS response message using Zenoh.
+rmw_ret_t
+rmw_send_response(const rmw_service_t * service,
+                  rmw_request_id_t * request_header,  // In parameter
+                  void * ros_response)
+{
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp",
+                         "[rmw_send_response] %s (%ld)",
+                         service->service_name,
+                         static_cast<rmw_service_data_t *>(service->data)->zn_response_topic_id_);
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_response, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    service,
+    service->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION
+  );
+
+  auto service_data = static_cast<rmw_service_data_t *>(service->data);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_data, RMW_RET_ERROR);
+
+  // ASSIGN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator =
+    &static_cast<rmw_service_data_t *>(service->data)->node_->context->options.allocator;
+
+  // SERIALIZE DATA ============================================================
+  size_t max_data_length = (
+    static_cast<rmw_service_data_t *>(service->data)
+      ->response_type_support_->getEstimatedSerializedSize(ros_response)
+  );
+
+  // Account for metadata
+  max_data_length += sizeof(request_header->sequence_number);
+
+  // Init serialized message byte array
+  char * response_bytes = static_cast<char *>(
+    allocator->allocate(max_data_length, allocator->state)
+  );
+
+  eprosima::fastcdr::FastBuffer fastbuffer(
+    response_bytes,
+    max_data_length
+  );  // Object that manages the raw buffer.
+
+  eprosima::fastcdr::Cdr ser(fastbuffer,
+                             eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
+                             eprosima::fastcdr::Cdr::DDS_CDR);  // Object that serializes the data.
+  if (!service_data->response_type_support_->serializeROSmessage(
+    ros_response, ser, service_data->response_type_support_impl_)
+  ) {
+    allocator->deallocate(response_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  size_t data_length = ser.getSerializedDataLength();
+
+  // ADD METADATA ==============================================================
+  // TODO(CH3): Again, refactor this into a modular set of functions eventually
+  size_t meta_length = sizeof(request_header->sequence_number);
+  memcpy(
+    &response_bytes[data_length], reinterpret_cast<char *>(&request_header->sequence_number), meta_length
+  );
+
+  // PUBLISH ON ZENOH MIDDLEWARE LAYER =========================================
+  zn_write_wrid(service_data->zn_session_,
```

Comment:
Is there a return code that can be checked here?

Reply:
Updated: 5452948477af25bf7c5d9a36c2fb81c024f61ae4